solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceBaseInputOperator.java

```diff
@@ -0,0 +1,319 @@
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    this.windowId = windowId;
+  }
+
+
+  protected abstract T convert(BytesXMLMessage message);
+
+  protected abstract void emitTuple(T tuple);
+
+  protected abstract Consumer getConsumer() throws JCSMPException;
+
+  protected abstract void clearConsumer() throws JCSMPException;
+
+  public void setProperties(JCSMPProperties properties)
+  {
+    this.properties = properties;
+
+  }
+
+
+  public IdempotentStorageManager getIdempotentStorageManager()
+  {
+    return idempotentStorageManager;
+  }
+
+  public void setUnackedMessageLimit(String unackedMessageLimit)
+  {
+    this.unackedMessageLimit = unackedMessageLimit;
+  }
+
+  public String getUnackedMessageLimit()
+  {
+    return unackedMessageLimit;
+  }
+
+  public void setConnectRetries(String connectRetries)
+  {
+    this.connectRetries = connectRetries;
+    logger.info("+++++++++++++++++++reconnectRetries: {}", this.connectRetries);
+  }
+
+  public void setReconnectRetries(String reconnectRetries)
+  {
+    this.reconnectRetries = reconnectRetries;
+    logger.info("+++++++++++++++++++reconnectRetries: {}", this.reconnectRetries);
+  }
+
+  public void setReapplySubscriptions(boolean state)
+  {
+    this.properties.setBooleanProperty(JCSMPProperties.REAPPLY_SUBSCRIPTIONS, state);
+  }
+
+  public void startConsumer()
+  {
+    try {
+      consumer = getConsumer();
+    } catch (JCSMPException e) {
+      DTThrowable.rethrow(e);
+    }
+  }
+
+
+  public class ReconnectCallbackHandler implements JCSMPReconnectEventHandler
+  {
+
+
+    @Override
+    public void postReconnect() throws JCSMPException
+    {
+
+      logger.info("++++++++++++++Solace client now Reconnected --  possibe Solace HA or DR fail-over +++++++++++++");
+      TCPDisconnected = false;
+
+    }
+
+    @Override
+    public boolean preReconnect() throws JCSMPException
+    {
+      DRFailover = false;
+      logger.info("++++++++++++++Solace client now in Pre Reconnect state -- possibe Solace HA or DR fail-over +++++++++++++");
+      TCPDisconnected = true;
+      return true;
+    }
+
+  }
+
+  public class PrintingSessionEventHandler implements SessionEventHandler
+  {
+
+
+    public void handleEvent(SessionEventArgs event)
+    {
+      logger.info("Received Session Event %s with info %s\n", event.getEvent(), event.getInfo());
+
+      // Received event possibly due to DR fail-ver complete
+      if (event.getEvent() == SessionEvent.VIRTUAL_ROUTER_NAME_CHANGED) {
+        DRFailover = true; // may or may not need recovery
+        TCPDisconnected = false;
+      }
+
+    }
+  }
+
+  public class CallbackMessageHandler implements XMLMessageListener
+  {
+
+    @Override
+    public void onException(JCSMPException e)
+    {
+      DTThrowable.rethrow(e);
+
+    }
+
+    @Override
+    public void onReceive(BytesXMLMessage msg)
+    {
+      arrivedTopicMessagesToProcess.add(msg);
```

Comment:
Couple of things:

- would put(msg) be better than add() since put is blocking and that is what you want when the queue is full (depends on the contract of XMLMessageListener)

- even with add() if it throws IllegalStateException (when the queue is full) who is going to handle it? Won't there be a loss of messages otherwise?

- don't we need to synchronize on the queue?

Reply:
It is a synchronized queue